Feature add index to session attempts

yoyama opened the pull request:
Add an index to resolve performance issue on session_attempts.

To support _create index concurrently_, re-write DatabaseMigrator.migrate() and support no transaction mode in Migration.

Original migrate() support two mode.
(a) Initial mode : all migrations in a transaction
(b) Upgrade mode : each migration run in separate transaction

New migrate() only support (b). Because of support no transaction mode.
In multiple digdag server cluster, if digdag servers boot simultaneously, migrate() may fail.
To avoid it, migrateWithRetry() is implemented.

muga commented on digdag-core/src/main/java/io/digdag/core/database/DatabaseMigrator.java:
```diff
@@ -76,78 +82,78 @@ public String getSchemaVersion()
-        RUN_ALL,
-        FINISHED,
-        MORE;
+        final int max_retry = 3;
```
minor: better to use camel case?

yoyama replied:
Oh! sorry. I fixed it.

yoyama commented on digdag-core/src/main/java/io/digdag/core/database/DatabaseMigrator.java:
```diff
@@ -156,9 +162,28 @@ private void createSchemaMigrationsTable(Handle handle, MigrationContext context
         handle.insert("insert into schema_migrations (name, created_at) values (?, now())", m.getVersion());
     }
+
+    public String getDatabaseType()
```
Thanks!. Fixed it.

yoyama commented on digdag-core/src/main/java/io/digdag/core/database/DatabaseMigrator.java:
```diff
@@ -76,78 +82,78 @@ public String getSchemaVersion()
+        logger.error("Database migration aborted.");
     }
 
     public void migrate()
```
I agree. Fixed it. Thanks!

yoyama commented on digdag-core/src/main/java/io/digdag/core/database/migrate/Migration.java:
```diff
@@ -17,5 +17,12 @@ default String getVersion()
+     * This is introduced because 'create index concurrently' cannot run in transaction.
+     * @return
+     */
+    default public boolean noTransaction() { return false; }
```
Thanks, I fixed it.

muga commented on digdag-core/src/main/java/io/digdag/core/database/migrate/Migration_20190318175338_AddIndexToSessionAttempts.java:
```diff
@@ -0,0 +1,23 @@
+        // DatabaseSessionStoreManager.getActiveAttemptCount uses this index.
+        if (context.isPostgres()) {
+            handle.update("create index concurrently session_attempts_on_site_id_and_state_flags_partial_2 on session_attempts"
+                    + " using btree(site_id) where state_flags & 2 = 0");
```
Can the update method wait for the sql completion? I just wanted to confirm it because it'd better that each migration is executed sequentially.

yoyama replied:
Yes. _create index concurrently_ also wait for the completion of sql.
Following is my experiment on above query to about 10,000,000 records.

```
digdag_test=# create index session_attempts_on_site_id_and_state_flags_partial_2 on session_attempts  using btree(site_id) where state_flag
s & 2 = 0;
CREATE INDEX
Time: 2458.866 ms (00:02.459)

digdag_test=# create index concurrently session_attempts_on_site_id_and_state_flags_partial_2 on session_attempts  using btree(site_id) whe
re state_flags & 2 = 0;
CREATE INDEX
Time: 3830.146 ms (00:03.830)
```
With _concurrently_, the response speed is late. So it waited for the completion.

muga replied:
Thanks! it makes sense 👍

yoyama commented:
@muga Thanks for your comment. I agree. I am preparing for document and _migrate_ command in this PR.

yoyama commented:
@muga Could you review again?

- Added _digdag migrate_ command.
  - _migrate_ command has two sub-command. _check_ and _run_
  - _check_ only confirms the list of migrations which should be applied.
  - _run_ execute migrations
  - migrate command directly connects to database. So it requires db connection info by -c or -o
- Added a documentation for migration.
